js/ui/workspaceMonitor.js

```diff
@@ -10,276 +10,144 @@ const WorkspaceMonitor = new Lang.Class({
+            // Hide in fullscreen mode
+            Main.overview.hide();
         }
     },
 
-    _windowGoingInvisibleCompleted: function() {
-        // Show the overview when the animation has ended.
-        this._visibleWindows -= 1;
-        this._updateOverview();
-    },
-
-    _destroyWindow: function(shellwm, actor) {
-        // _destroyWindow is not called for minimized windows
-        // so if the window is in _knownWindows then we handle it
-        if (!this._windowIsKnown(actor.meta_window)) {
-            return;
-        }
-
-        this._windowGoingInvisible();
-
-        // We'll show the overview when the destroy animation ends
-        this._addDestroyedWindow(actor.meta_window);
+    _trackApp: function(app) {
+        if (!(app in this._trackedApps))
+            this._trackedApps[app] = app.connect('windows-changed', Lang.bind(this, this._onAppWindowsChanged));
     },
 
-    _destroyWindowCompleted: function(shellwm, actor) {
-        if (this._windowIsDestroyed(actor.meta_window)) {
-            this._windowGoingInvisibleCompleted();
-            this._removeDestroyedWindow(actor.meta_window);
+    _untrackApp: function(app) {
+        if (app in this._trackedApps) {
+            app.disconnect(this._trackedApps[app]);
+            delete this._trackedApps[app];
         }
     },
 
-    _minimizeWindow: function(shellwm, actor) {
-        if (!this._windowIsKnown(actor.meta_window)) {
-            return;
-        }
-
-        this._windowGoingInvisible();
-
-        // We'll show the overview when the destroy animation ends
-        this._addMinimizedWindow(actor.meta_window);
+    _setAppVisible: function(app, visible) {
+        if (visible)
+            this._visibleApps.add(app);
+        else
+            this._visibleApps.delete(app);
     },
 
-    _minimizeWindowCompleted: function(shellwm, actor) {
-        if (this._windowIsMinimized(actor.meta_window)) {
-            this._windowGoingInvisibleCompleted();
-            this._removeMinimizedWindow(actor.meta_window);
+    _onAppStateChange: function(appSystem, app) {
+        if (app.get_state() == Shell.AppState.RUNNING) {
+            this._trackApp(app);
         }
     },
 
-    _realMapWindow: function(metaWindow) {
-        // If we don't know about it then we don't handle it
-        if (!this._windowIsKnown(metaWindow)) {
-            return;
-        }
+    _onAppWindowsChanged: function(app) {
+        this._setAppVisible(app, this._appHasVisibleWindows(app));
 
-        // If the window was minimized remove it
-        if (this._windowIsMinimized(metaWindow)) {
-            this._removeMinimizedWindow(metaWindow);
+        // We need to track the stopped state here (instead of in _onAppStateChange),
+        // and independently from setting an app visible, because there are apps
+        // that go into "running" -> "stopped" -> "running" states because of the
+        // way they show their windows. This is the case of the Maps (marble) app.
```

Comment:
The following code block implies that it's possible for an app to end up in a `STOPPED` state with one window still being visible. I think that is a bug in the `ShellWindowTracker`/`ShellAppSystem`/`ShellApp` machinery. It should never be possible that the app is stopped but a window is still visible, and we should not work that around here.